### ISSUE TITLE
IsEmpty() should check for `.empty()`, not `.size() == 0`

### ISSUE THREAD

**Issue by nabijaczleweli** [banditcpp/snowhouse#14]

**Comment by nabijaczleweli** [comment]
Any update on this?

**Comment by joakimkarlsson** [comment]
Thanks for reporting!

Is there a practical consequence to this?

**Comment by nabijaczleweli** [comment]
Yes, inability to check for "emptiness" of things that are not containers (custom smart pointers or wrappers, for example).

**Comment by joakimkarlsson** [comment]
Yes, that makes sense. Thanks.

Right now, IsEmpty() is implemented as a special case of HasLength(). We need to promote IsEmpty() to using its own IsEmptyConstraint.

**Comment by nabijaczleweli** [comment]
Any update on this?

**Comment by nabijaczleweli** [comment]
Any update on this?

**Comment by sbeyer** [comment]
I just looked into it in the hope that it would be simple.

For example, C++11's `std::forward_list` has no `size()` method. I included it in sequence_container_tests.cpp and noticed that a lot more than `isEmpty()` uses the `size()` method, for example, also the `All` operator does. Interestingly, also the so-far compiling tests fail on `std::forward_list`... Very weird. I pushed the work in progress to the [more-container-tests branch](/banditcpp/snowhouse/tree/more-container-tests), if interested, but I am not going to work on it now.
